ulib/FStar.Fin.fst

```diff
@@ -40,44 +49,169 @@ let seqn (n: nat) (a: Type) = s: S.seq a {S.length s = n}
 let in_ (#a: Type) (s: S.seq a) = n: nat{n < S.length s}
 
 (** Find an index of an element in [s] starting from [i] that validates [p]  *)
-let rec find (#a: Type) (s: S.seq a) (p: (a -> bool)) (i: in_ s)
+let rec find (#a: Type) (s: S.seq a) (p: (a -> bool)) (i: under (S.length s))
     : Pure (option (in_ s))
       (requires True)
-      (ensures
-        (function
-          | None -> (forall (k: nat{i <= k /\ k < S.length s}). p (S.index s k) == false)
-          | Some j -> i <= j /\ p (S.index s j)))
+      (ensures (function
+                | None -> (forall (k: nat{i <= k /\ k < S.length s}). p (S.index s k) == false)
+                | Some j -> i <= j /\ p (S.index s j)))
       (decreases (S.length s - i)) =
-  if p (S.index s i) then Some i else if i + 1 < S.length s then find #a s p (i + 1) else None
+  if p (S.index s i) 
+  then Some i 
+  else if i + 1 < S.length s 
+       then find #a s p (i + 1) 
+       else None
 
 (** Given a sequence [s] all of whose elements are at most [n], if the
     length of [s] is greater than [n], then there are two distinct
     indexes in [s] that contain the same element *)
-let rec pigeonhole (#n: nat) (s: S.seq (fin n))
+let rec pigeonhole (#n: pos) (s: S.seq (under n))
     : Pure (in_ s * in_ s)
       (requires S.length s = n + 1)
       (ensures (fun (i1, i2) -> i1 < i2 /\ S.index s i1 = S.index s i2))
-      (decreases n) =
-  if n = 0
-  then (match S.index s 0 with )
-  else
-    if n = 1
-    then
-      (assert (S.index s 0 = S.index s 1);
-        0, 1)
-    else
-      let k0 = S.index s 0 in
-      match find s (fun k -> k = k0) 1 with
-      | Some i -> 0, i
-      | None ->
-        let reduced_s:S.seq (fin (n - 1)) =
-          S.init #(fin (n - 1))
-            n
-            (fun i ->
-                let k:nat = S.index s (i + 1) in
-                assert (k <> k0);
-                if k < k0 then (k <: fin (n - 1)) else (k - 1 <: fin (n - 1)))
-        in
-        let i1, i2 = pigeonhole reduced_s in
-        i1 + 1, i2 + 1
+      (decreases n) = 
+   if n = 1 then (0, 1)
+   else let k0 = S.index s 0 in
+     match find s (fun k -> k = k0) 1 with
+     | Some i -> 0, i
+     | None -> let (i1,i2) = 
+                pigeonhole (S.init #(under (n-1)) n
+                                   (fun i -> let k = S.index s (i+1) in
+                                          if k<k0 then k else k-1))
+              in (i1+1, i2+1)
+   
+(** Here we prepare to prove pigeonhole principle for a finite sequence
+    with a custom equivalence relation (as opposed to eqtype).
+    
+    Think setoids. *)
+
+(** Following code is extracted from CuteCAS, which will eventually make 
+    its way into F* -- when I wrap things up with most important notions
+    of abstract algebra. 
+    
+    As I port more code from my CAS project to F*, such things will be
+    moved to separate modules. -- Alex Rozanov *)
+ 
+type binary_relation (a: Type) = a -> a -> bool
```

Comment:
Hi @hacklex, some of these definitions are already there in `ulib/FStar.Preorder.fst`.  Can we use them instead of redefining? They are not opaque, but if your proof goes through with them, it would be nice to reuse.

Reply:
Hi @aseemr!

I took a look. 
We could reuse these here, yes.

I advise to make those opaque though, I managed to catch performance regressions from non-opaque forall-definitions even in smaller lemmas than those I wrote for pigeonhole principle proof. This really reduces verification times greatly.

Also, now that I checked out FStar.Preorder, I think we should structurally expand relations like this:
FStar.Relations
  + binary_relation
  + symmetry, transitivity, reflexivity, antisymmetry, antireflexivity
  + equivalence
  + preorder
  + total_order

and further develop it with custom equivalence in mind, for example, 
```fst
let preorder #t (eq: equivalence t) = rel: binary_relation t { reflexive rel /\ transitive rel }
```

or alternatively keep the parametrized versions in a separate module.

If you want, I can provide a more general version for binary relations. I'm not sure though, how much code depends on the current FStar.Preorder, and should we rewire it or not depends on how much will break from it.

But adding a few `reveal_opaque`s where it is really needed sounds like a very small price to me, as sometimes it offers good gains in verification speed. 

So if you're interested, I can write a more general module for binary relations, and then we slowly migrate to a more mature, in architectural sense, module. 

For this PR, I'll just add the `fsti` as you asked, and reuse the current FStar.Preorder.